src/compiler/checker.ts

```diff
@@ -24924,16 +24937,102 @@ namespace ts {
+        }
+
+        function combineIntersectionThisParam(left: Symbol | undefined, right: Symbol | undefined, mapper: TypeMapper | undefined): Symbol | undefined {
+            if (!left || !right) {
+                return left || right;
+            }
+            // A signature `this` type might be a read or a write position... It's very possible that it should be invariant
+            // and we should refuse to merge signatures if there are `this` types and they do not match. However, so as to be
+            // pessimistic when contextual typing, for now, we'll union the `this` types.
+            const thisType = getUnionType([getTypeOfSymbol(left), instantiateType(getTypeOfSymbol(right), mapper)]);
+            return createSymbolWithType(left, thisType);
+        }
+
+        function combineIntersectionParameters(left: Signature, right: Signature, mapper: TypeMapper | undefined) {
+            const leftCount = getParameterCount(left);
+            const rightCount = getParameterCount(right);
+            const longest = leftCount >= rightCount ? left : right;
+            const shorter = longest === left ? right : left;
+            const longestCount = longest === left ? leftCount : rightCount;
+            const eitherHasEffectiveRest = (hasEffectiveRestParameter(left) || hasEffectiveRestParameter(right));
+            const needsExtraRestElement = eitherHasEffectiveRest && !hasEffectiveRestParameter(longest);
+            const params = new Array<Symbol>(longestCount + (needsExtraRestElement ? 1 : 0));
+            for (let i = 0; i < longestCount; i++) {
+                let longestParamType = tryGetTypeAtPosition(longest, i)!;
+                if (longest === right) {
+                    longestParamType = instantiateType(longestParamType, mapper);
+                }
+                let shorterParamType = tryGetTypeAtPosition(shorter, i) || unknownType;
+                if (shorter === right) {
+                    shorterParamType = instantiateType(shorterParamType, mapper);
+                }
+                const unionParamType = getUnionType([longestParamType, shorterParamType]);
+                const isRestParam = eitherHasEffectiveRest && !needsExtraRestElement && i === (longestCount - 1);
+                const isOptional = i >= getMinArgumentCount(longest) && i >= getMinArgumentCount(shorter);
+                const leftName = i >= leftCount ? undefined : getParameterNameAtPosition(left, i);
+                const rightName = i >= rightCount ? undefined : getParameterNameAtPosition(right, i);
+
+                const paramName = leftName === rightName ? leftName :
+                    !leftName ? rightName :
+                    !rightName ? leftName :
+                    undefined;

```

Comment:
uhhhhhhhhh, why not just `leftName || rightName`? I'm not even sure it's observable, but even if it is, some name is better than `arg0`.

Reply:
Because that's what we merged for `combineUnionParameters` in https://github.com/microsoft/TypeScript/pull/32056 :V

Also, "some name is better than arg0" isn't really true. Names carry semantic meaning, so let's say you have `(index: number) => any` and `(object: Whatever): any ` - calling the result `index` where's it's type is `number | Whatever` has the potential to be misleading. Particularly problematic is when the types are the _same_, but semantically different, like
`(length: number): any[]` and `(firstElement: number): any[]` - yeah, `number` is the type of both, but there're not semantically the same! So calling both "length" or "firstElement" would be dead wrong. Now, I had a _super old_ version of this that concatenated the names in cases like this, to `lengthOrFirstElement`, but that was shot down as too easily producing names that are unwieldy as unions grow. And thus, the preference for `arg0` was born. It carries no information, thus you can draw no (potentially inaccurate) conclusion as to the usage of the parameter.

Reply:
If @dragomirtitian cares enough to fix it, that's good enough for me. =)